WIP: Add fractional attribution accelerator code and markdown (#2)
snowplow/marketing-attribution-accelerator: content/modeling/modeling_3.md

For the sake of simplicity we have selected the variables that you will most likely need to overwrite, the rest can be changed at a later stage if and when it is needed.

- `conversion_window_start_date`: The start date in UTC for the window of conversions to include
- `conversion_window_end_date`: The end date in UTC for the window of conversions to include
- `conversion_hosts`: url_hosts to consider
- `path_lookback_steps`: The limit for the number of marketing channels to look at before the conversion (default is 0 = unlimited)
- `path_lookback_days`: Restrict the model to marketing channels within this many days of the conversion (values of 30, 14 or 7 are recommended)
- `path_transforms`: An array of path transforms and their arguments (see below section **Path Transform Options**)
- `consider_intrasession_channels`: Boolean. If false, only considers the channel at the start of the session (i.e. first page view). If true, considers multiple channels in the conversion session as well as historically.

The default source schemas and tables used by the snowplow_fractribution package are:

> - *atomic_derived.snowplow_web_page_views* for the page_views data (page_views_source)

Author: Ah thanks - fixed.

- *atomic.events* for the Snowplow event data (conversions_source)

If either of these differ in your warehouse, set the correct names as variables in your own dbt_project.yml, e.g.:
-  `page_views_source`: `custom_schema_derived.snowplow_web_page_views`

> -  `conversions_source`: `custom_schema_derived.snowplow_ecommerce_transaction_interactions`

Reviewer: ```suggestion
-  `conversions_source`: `custom_schema_derived.conversions`
```

Reviewer: Just realised the ecommerce model will only support BigQuery at first. Let's change these only once it becomes available for Snowflake, too

Author: Ah. So does this mean that I should change the tracking steps and all references relating to the new Ecommerce plugin back to what I had originally?

Reviewer: The tracking bit for Ecommerce should all point at the "old" ecommerce plugin as the new one is not yet released (and even if it gets released in a few days unfortunately we have to wait probably till mid-January until we get Snowflake support, or till the Fractribution dbt package starts supporting BigQuery). Peter had a look and said currently the transaction API bit is from the new Ecommerce one, but the plugin is from the old one, therefore the transaction API part needs to be changed. He said he will kindly make that change (probably tomorrow).

Author: Yeah - I changed some references from the old one to the new when Peter recommended the new Ecommerce model. If he wants to make the changes, that's great! The TODOs I left in there in regards to the new Ecommerce plugin name uncertainties I had will need to be removed also.

Reviewer: This is done now, too, Peter has revised it!

> For example, if you do not wish the package to query your `atomic.events` table to get the conversion value data, the `conversions_source` variable can point to a different table, such as the derived table `snowplow_ecommerce_transaction_interactions` created by the ecommerce model. **(TODO: Add link)**

Reviewer: ```suggestion
For example, one option would be to create your own snowplow-web incremental data model for transaction/conversion events, and add a reference to that model as the conversions_source variable, e.g. `conversions_source: {{ ref('custom_conversions_model') }}`
```

> Another option would be to create your own incremental data model for transaction/conversion events, and add a reference to that model as the conversions_source variable, e.g. `conversions_source: {{ ref('custom_conversions_model') }}`

Reviewer: ```suggestion

```

```yml
vars:
  conversion_window_start_date: '2022-06-03'
  conversion_window_end_date: '2022-08-01'
  conversion_hosts: ['poplindata.com']
  path_lookback_steps: 0
  path_lookback_days: 30
  path_transforms: [['Exposure', null]]
  consider_intrasession_channels: false
  conversions_source: 'atomic.sample_events_fractribution'
```

**Path Transform Options**

Paths to conversion are often similar, but not identical. As such, path transforms reduce unnecessary complexity in similar paths before running the attribution algorithm.

There are five path transform options available:
- Exposure (default in this package): the same events in succession are reduced to one: A -> A -> B becomes A -> B. A compromise between first and unique.
- Unique: all events in a path are treated as unique (no reduction of complexity). Best for smaller datasets (small lookback window) without a lot of retargeting.
- First: keep only the first occurrence of any event: A -> B -> A becomes A -> B. Best for brand awareness marketing.
- Frequency: keep a count of the events' frequency: A -> A -> B becomes A(2) -> B. Best when there is a lot of retargeting.

> - Recency: time bucket events depending on when they occurred before the conversion and collapse those in the same bucket (square braces indicate minutes before conversion event): A[15-30] -> B[1-5] -> A[1-5] -> B[1-5] becomes A[15-30] -> B[1-5] -> A[1-5]. Best for larger lookback windows.

Reviewer: I have noticed `Recency` is not among UDFs, is it missing from the model or it should be removed from here? Also not sure if the example is correct here as there are 2 `A` groups and if I understand correctly it should be unique after collapsing (?).

Author: Oh good catch, thank you - I shouldn't have added this. Deleted. The example is a little confusing anyway - the Bs are collapsed because they are in the same bucket [1-5] but the As aren't because they are different time buckets [15-30] vs [1-5]

***

#### **Step 5:** Explore the data created by your dbt models

Take some time to familiarise yourself with the derived tables. These tables are used in the next step to fractionally attribute revenue to channels. Tables output by the snowplow_fractribution dbt package are:

- `snowplow_fractribution_channel_counts`: Number of events grouped by channel, campaign, source and medium.
- `snowplow_fractribution_channel_spend`: Spend on each channel, used in ROAS calculations.
- `snowplow_fractribution_conversions_by_customer_id`: Conversion revenue for each conversion, along with the associated customerid.
- `snowplow_fractribution_path_summary`: Summary of different path combinations and associated conversion/non-conversions.
- `snowplow_fractribution_paths_to_conversion`: Path combinations leading to conversion.
- `snowplow_fractribution_paths_to_non_conversion`: Path combinations leading to non-conversion.
- `snowplow_fractribution_sessions_by_customer_id`: Channel information by session timestamp, where an event timestamp is considered as the session start.

***

> Next we will run a Python script to create the actual Fractribution table.

Reviewer: I have tried to run all of the attribution models and two of them won't run due to the same error (first_touch, last_touch): `TypeError: field Referral: Cannot merge type <class 'snowflake.snowpark.types.LongType'> and <class 'snowflake.snowpark.types.FloatType'>` It happens when it tries to create a snowpack dataframe. I have only just discovered it so I do not have time to investigate further today, but hopefully you can recreate the issue as I was using the test data . If not I will have a look tomorrow.

Author: Thanks, I missed this. Mike and I fixed this today, should work as expected now.

**Configure the default channel_classification macro**

The channel_classification macro is used to perform channel classifications. This can be altered to generate your expected channels if they differ from the channels generated in the default macro. It is highly recommended that you examine and configure this macro when using your own data, as the default values will not consider any custom marketing parameters.

If you wish to change the channel classification macro, copy the `channel_classification.sql` file from the macros folder in the snowplow_fractribution package (at `[dbt_project_name]/dbt_packages/snowplow_fractribution/macros/channel_classification.sql`) and add it to the macros folder of your own dbt project. Update the SQL and save the file.
 
 
If you have added one or more of these macros to your own project's macros folder, the next step is to add the following to your dbt_project.yml file (replacing `your_project_name` with the name of your project, found at the top of the dbt_project.yml file after `name:`):

```yml
dispatch:

>   - macro_namespace: snowplow_fractribution

Reviewer: I think this is not necessary as your own project takes precedence by default. When I tested it locally it did overwrite the original configurations without explicitly adding this.

Author: okay great, thanks! I'll remove.
